File: NCP Browser V2/chrome/runtime/NativeBrowserScripting.js
// Base Runtime
if (!chrome.runtime) {
    chrome.runtime = {};
    chrome.runtime.id = ncp_runtime.id;
}

// On Connect and handlers/events
if (!chrome.runtime.onConnect) {
    chrome.runtime.onConnect = {};
    chrome.runtime.onConnect.listeners = new Array();

    chrome.runtime.onConnect.addListener = function (method) {
        chrome.runtime.onConnect.listeners.push(method);
    }

    chrome.runtime.onConnect.fire = function (port) {
        for (var i = 0; i < chrome.runtime.onConnect.listeners.length; i++) {
            chrome.runtime.onConnect.listeners[i](JSON.parse(port));
        }
    }
}

// On Connect External handlers/events
if (!chrome.runtime.onConnectExternal) {
    chrome.runtime.onConnectExternal = {};
    chrome.runtime.onConnectExternal.listeners = new Array();

    chrome.runtime.onConnectExternal.addListener = function (method) {
        chrome.runtime.onConnectExternal.listeners.push(method);
    }

    chrome.runtime.onConnectExternal.fire = function (port, channelId) {
        for (var i = 0; i < chrome.runtime.onConnectExternal.listeners.length; i++) {
            port = JSON.parse(port);
            port.channelId = channelId;
            port.onMessage = {};
            port.onMessage.listeners = new Array();
            port.onMessage.addListener = function (method) {
                this.listeners.push(method);
            }
            port.onMessage.fire = function (message, port) {
                for (var i = 0; i < this.listeners.length; i++) {                    
                    this.listeners[i](message, port);
                }
            }

            port.onDisconnect = {};
            port.onDisconnect.listeners = new Array();
            port.onDisconnect.addListener = function (method) {
                this.listeners.push(method);
            }
            port.onDisconnect.fire = function (message) {
                for (var i = 0; i < this.listeners.length; i++) {
                    this.listeners[i](message);
                }
            }

            port.postMessage = function (message) {
                ncp_runtime.receiveMessage(this.channelId, JSON.stringify(message));
            }

            chrome.runtime.port.openPorts.push(port)

            chrome.runtime.onConnectExternal.listeners[i](port);
        }
    }
}

if (!chrome.runtime.port) {
    chrome.runtime.port = {};
    chrome.runtime.port.openPorts = new Array();
    chrome.runtime.onPortMessage = {};
    chrome.runtime.onPortMessage.fire = function (message, channelId) {
        message = JSON.parse(message);
        for (var i = 0; i < chrome.runtime.port.openPorts.length; i++) {
            if (chrome.runtime.port.openPorts[i].channelId == channelId) {
                chrome.runtime.port.openPorts[i].onMessage.fire(message, chrome.runtime.port.openPorts[i]);
            }
        }
    }
}

if (!chrome.runtime.nativePort) {
    chrome.runtime.nativePort = {};
    chrome.runtime.nativePort.openPorts = new Array();
}

if(!chrome.runtime.onNativeMessage) {
    chrome.runtime.onNativeMessage = {};
    chrome.runtime.onNativeMessage.fire = function(message) {
        // Send message to all open ports
        for (var i = 0; i < chrome.runtime.onMessage.listeners.length; i++) {
            chrome.runtime.onMessage.fire(message);
        }
    }
}



// Connect Native method + handlers/events
if (!chrome.runtime.connectNative)
{
    chrome.runtime.connectNative = function (x) {

        if (!chrome.runtime.onMessage) {
            chrome.runtime.onMessage = {};
            chrome.runtime.onMessage.listeners = new Array();
            chrome.runtime.onMessage.fire = function (data) {
                //if (portId == chrome.runtime.id) {
                    for (var i = 0; i < chrome.runtime.onMessage.listeners.length; i++) {
                        chrome.runtime.onMessage.listeners[i](data);
                    }
                //}
            }
        }

        if (!chrome.runtime.onDisconnect) {
            chrome.runtime.onDisconnect = {};
            chrome.runtime.onDisconnect.listeners = new Array();
            chrome.runtime.onDisconnect.fire = function (data) {
                //if (portId == chrome.runtime.id) {
                    for (var i = 0; i < chrome.runtime.onDisconnect.listeners.length; i++) {
                        chrome.runtime.onDisconnect.listeners[i](data);
                    }
                //}
            }
        }

        port = ncp_runtime.connectNative(x,window.location.hostname);
        returnPort = {
            name: port.name,
            disconnect: function () {
                // TODO
                chrome.runtime.onDisconnect.fire(this);
            },
            onDisconnect: {
                addListener: function (method) {
                    chrome.runtime.onDisconnect.listeners.push(method);

                    //TODO?
                    //ncp_runtime.onMessageListeners(port.name, chrome.runtime.onMessage.listeners.length);
                }
            },
            onMessage: {
                addListener: function (method) {
                    chrome.runtime.onMessage.listeners.push(method);
                    //ncp_runtime.onMessageListeners(port.name, chrome.runtime.onMessage.listeners.length);
                }
            },
            postMessage: function (message) {
                //delete message.ciscoChannelProperties;
                //message.client = this.client;
                //message = {
                //    ciscoChannelMessage: message
                //};
                message = JSON.stringify(message);
                ncp_runtime.postMessage(this.name, message);
            }            
        };
        return returnPort;
    };
}

// Base Management
if (!chrome.management) {
    chrome.management = {};
    chrome.management.asynchronousCallbacks = new Array();
}

// Management Get
if (!chrome.management.get)
{
    // This method is async
    chrome.management.get = function (extensionId, callback) {
        // build a callback id to use as a key for the callback list
        var d = new Date();
        var callbackId = extensionId + d.getTime();

        // add callback to list
        chrome.management.asynchronousCallbacks[callbackId] = callback;

        // call async browserscripting method
        ncp_runtime.management_get(extensionId, callbackId);
    }
}

// Management Get CallBack
if (!chrome.management.getCallback)
{
    // This method is an async callback
    chrome.management.getCallback = function (extensionInfo, callbackId) {
        if(chrome.management.asynchronousCallbacks(callbackId))
        {
            // get a reference to the method
            var cb = chrome.management.asynchronousCallbacks[callbackId];

            // delete method from callback list
            delete chrome.management.asynchronousCallbacks[callbackId];

            // call method
            cb(JSON.parse(extensionInfo));
        }
    }
}

// Extension
if (!chrome.extension) {
    chrome.extension = {}
}

if (!chrome.extension.getURL) {
    chrome.extension.getURL = function (url) {
        return ncp_runtime.extension_getURL(url);
    }
}

File: NCP Browser V2/chrome/runtime/NativeBrowserScripting.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace NCP_Browser.chrome.runtime
{
    [System.Runtime.InteropServices.ComVisibleAttribute(true)]
    public class NativeBrowserScripting
    {
        internal static string Name = "ncp_runtime";
        private string _id;

        public static List<ChromeFrame> ActiveOnMessageListeners = new List<ChromeFrame>();
        private CefSharp.WinForms.ChromiumWebBrowser chromiumWebBrowser;
        private CefSharp.WinForms.ChromiumWebBrowser browser;
        private BrowserTabUserControl browserTabUserControl;
        private Salesforce salesforce;
        private object loadLock;
        private bool doneInit = false;


        public NativeBrowserScripting(CefSharp.WinForms.ChromiumWebBrowser browser, BrowserTabUserControl browserTabUserControl, String Id, Salesforce salesforce)
        {
            // TODO: Complete member initialization
            this.browser = browser;
            this.browserTabUserControl = browserTabUserControl;
            this._id = Id;
            this.salesforce = salesforce;
            this.loadLock = new object();

        }

        public string id
        {
            get
            {
                return this._id;
            }
        }

        public Port connectNative(string application, string hostname)
        {
            if (application == "com.cisco.jabber.jsdk")
            { 
                if(Salesforce.NativeMessagingExtensions.Values.Where(x => x.Name == "ppbllmlcmhfnfflbkbinnhacecaankdh").Count() == 1)
                {
                    Port p = new Port();
                    p.name = application;

                    var ex = Salesforce.NativeMessagingExtensions.Values.Where(x => x.Name == "ppbllmlcmhfnfflbkbinnhacecaankdh").First();
                    ex.connectNative(hostname, @"C:\Program Files (x86)\Cisco Systems\Web Communicator\CiscoWebCommunicatorAddOn.exe");

                    return p;
                }
                else
                {
                    return null;
                }
            }
            else
            {
                return null;
            }
            
        }

        public void postMessage(string application, string message)
        {
            string portId = String.Empty;
            if(application == "com.cisco.jabber.jsdk")
            {
                portId = "ppbllmlcmhfnfflbkbinnhacecaankdh";
            }
            
            if (Salesforce.NativeMessagingExtensions.Values.Where(x => x.Name == portId).Count() == 1)
            {
                Salesforce.NativeMessagingExtensions.Values.Where(x => x.Name == portId).First().WriteText(message);
                return;
            }
            else
            {
                return;
            }
        }

        public void receiveMessage(string channelid, string message)
        {            
            //salesforce.browserTabs.Values.ToList().ForEach(x =>
            //{
            //    x.QueueJavascriptToAllFrames(String.Format("chrome.runtime.onMessage.fire('{0}');", message.Replace("\"", "\\\"")));
            //});

            BrowserScripting.MessageCallbacks.Where(x => x.Channel == channelid).First().JavacriptCallback.ExecuteAsync(new object[] { message });
        }

        public void onMessageListeners(string portId, int ActiveListeners)
        {
            var listen = NativeBrowserScripting.ActiveOnMessageListeners.Where(x => x.portId == portId).FirstOrDefault();
            if(listen == null)
            {
                listen = new ChromeFrame();
                listen.portId = portId;
                listen.browser = this.chromiumWebBrowser;
                NativeBrowserScripting.ActiveOnMessageListeners.Add(listen);
            }
            listen.ActiveListeners = ActiveListeners;
        }

        public void management_get(string extensionId, string callbackId)
        {
            management_getReturn(extensionId, callbackId);
            return;
        }



        public async Task management_getReturn(string extensionid, string callbackId)
        {
            await Task.Run(() =>
            {
                ExtensionInfo extensionInfo = new ExtensionInfo();
                extensionInfo.id = extensionid;
                if(extensionid == "ppbllmlcmhfnfflbkbinnhacecaankdh")
                {
                    extensionInfo.description = "Enables Cisco Jabber phone calls and video calls";
                    extensionInfo.enabled = true;
                    extensionInfo.homepageUrl = "https://chrome.google.com/webstore/detail/ppbllmlcmhfnfflbkbinnhacecaankdh";
                    extensionInfo.hostPermissions = new string[0];
                    extensionInfo.icons = new IconInfo[0];
                    extensionInfo.installType = "admin";
                    extensionInfo.isApp = false;
                    extensionInfo.mayDisable = true;
                    extensionInfo.name = "Cisco Web Communicator";
                    extensionInfo.offlineEnabled = false;
                    extensionInfo.optionsUrl = "";
                    extensionInfo.permissions = new string[2];
                    extensionInfo.permissions[0] = "management";
                    extensionInfo.permissions[1] = "nativeMessaging";
                    extensionInfo.shortName = "Cisco Web Communicator";
                    extensionInfo.type = "extension";
                    extensionInfo.updateUrl = "https://clients2.google.com/service/update2/crx";
                    extensionInfo.version = "3.1.0.363";

                }

                //chromiumWebBrowser.GetBrowser().MainFrame.ExecuteJavaScriptAsync(String.Format("chrome.runtime.management.getCallback('{0}','{1}');", Newtonsoft.Json.JsonConvert.SerializeObject(extensionInfo, Newtonsoft.Json.Formatting.Indented, new Newtonsoft.Json.JsonSerializerSettings() { StringEscapeHandling = Newtonsoft.Json.StringEscapeHandling.Default }), callbackId));
            });
            return;
        }

        public string extension_getURL(string url)
        {
            return String.Format("cefsharp-extension://{0}/{1}", this._id, url);
        }

        public void DoneInitializing()
        {
            // I had to add a threadlock here and a boolean value to signify that has been called.
            // For some reason the chromium javascript runtime was doubling the invocations
            lock(loadLock)
            {
                if(!doneInit)
                {
                    this.browser.GetBrowser().MainFrame.ExecuteJavaScriptAsync(NCP_Browser.Properties.Resources.background.Replace("chrome-extension", "cefsharp-extension"));
                    
                    this.browserTabUserControl.DoneInitializing();

                    doneInit = true;
                }
            }
            
        }

        public class ExtensionInfo
        {
            public string description { get; set; }
            public bool enabled { get; set; }
            public string homepageUrl { get; set; }
            public string[] hostPermissions { get; set; }
            public IconInfo[] icons { get; set; }
            public string id { get; set; }
            public string installType { get; set; }
            public bool isApp { get; set; }
            public bool mayDisable { get; set; }
            public string name { get; set; }
            public bool offlineEnabled { get; set; }
            public string optionsUrl { get; set; }
            public string[] permissions { get; set; }
            public string shortName { get; set; }
            public string type { get; set; }
            public string updateUrl { get; set; }
            public string version { get; set; }
        }

        public class IconInfo
        {
            public int size { get; set; }
            public string url { get; set; }
        }
        
        public object Disconnect(object x, object[] x2)
        {
            return null;
        }

        public class ChromeFrame
        {
            public string portId { get; set; }
            public int ActiveListeners { get; set; }
            public CefSharp.WinForms.ChromiumWebBrowser browser { get; set; }
        }
    }
}
